### PR TITLE
Stream reuse

### LINKED ISSUE #47: Reduce number of TLS connections to forwarded (DoT) when using "forward-tls-upstream"
Hi.
First of all I want to thank you very much for your work about unbound!!
I mainly use unbound as forwarder to Cloudflare (1.1.1.1 and 1.0.0.1). So unbound listen on my Linux on port 53 and forward DNS queries to 1.1.1.1:853 via TLS (DNS over TLS). I use "forward-tls-upstream" option.
I have notice that unbound, for each new lookup, create a new connection to 1.1.1.1 on port 853 in order to process the lookup. This is tested also thanks to netstat command.
Well, on a performance point of view, I expect that if unbound will go to establish only one (or much less I mean) TLS connection to 1.1.1.1:853 (and leave it UP) the DNS queries will be processed more fastly because unbound don't need time to establish a new TLS connection each time to 1.1.1.1/1.0.0.1.
Is there the possiblity to set unbound to use one and always the same connection to 1.1.1.1:853 for all DNS over TLS queries?
If not, is there a plan to add an option to unbound in order to set what I'm asking?
Many thanks for you time.

This is the output of "netstat" command that show how many connections unbound do to DoT forwarder (Cloudflare in this case). In this example only one client is connected to unbound DNS Server (port 53):

```
root@server:~# netstat -anp | grep -i 53
tcp        0      0 0.0.0.0:53              0.0.0.0:*               LISTEN     620/unbound
tcp        0      0 192.168.1.10:40436     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48402     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48422     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40402     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48432     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48428     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40408     1.0.0.1:853             TIME_WAIT   -
tcp      225    137 192.168.1.10:48444     1.1.1.1:853             ESTABLISHED 620/unbound
tcp      152      0 192.168.1.10:40456     1.0.0.1:853             ESTABLISHED 620/unbound
tcp        0      0 192.168.1.10:48426     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40410     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48400     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40434     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40404     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40412     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48396     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48430     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48410     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40470     1.0.0.1:853             ESTABLISHED 620/unbound
tcp      377      0 192.168.1.10:48442     1.1.1.1:853             ESTABLISHED 620/unbound
tcp        0      0 192.168.1.10:48420     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40414     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40472     1.0.0.1:853             ESTABLISHED 620/unbound
tcp        0      0 192.168.1.10:40446     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48418     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48416     1.1.1.1:853             TIME_WAIT   -
tcp      385      0 192.168.1.10:48446     1.1.1.1:853             ESTABLISHED 620/unbound
tcp      152      0 192.168.1.10:40460     1.0.0.1:853             ESTABLISHED 620/unbound
tcp        0      0 192.168.1.10:40426     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40420     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:48436     1.1.1.1:853             TIME_WAIT   -
tcp      377      0 192.168.1.10:48440     1.1.1.1:853             ESTABLISHED 620/unbound
tcp        0      0 192.168.1.10:48408     1.1.1.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40428     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40416     1.0.0.1:853             TIME_WAIT   -
tcp        0      0 192.168.1.10:40406     1.0.0.1:853             TIME_WAIT   -
udp     6016      0 0.0.0.0:53              0.0.0.0:*                          620/unbound
```

### PR DESCRIPTION
**Pull request by wcawijngaards**
This implements stream reuse towards Unbound's upstream servers.  When a query is sent to a server for which there is an open socket, the socket is reused without opening a new stream to the server.  For TCP and TLS queries, for IPv4 and IPv6.  Sockets are opened per-thread, since the per-thread worker keeps track of the reused sockets.

### CODE REVIEW

**Comment by wcawijngaards** [comment]
This PR fixes issue #47

**Comment by pyropeter** [comment]
Will this be part of 1.12.0 or are you in feature freeze?

**Comment by wcawijngaards** [comment]
This is not in 1.12.0, because we are in feature freeze for it.  It is waiting for review, and maybe some fixes after that.

**Comment by djcater** [comment]
Will this cover both upstream DNS-Over-TLS connections and upstream DNS-Over-HTTPS connections?

**Comment by ralphdolmans** [comment]
> Will this cover both upstream DNS-Over-TLS connections and upstream DNS-Over-HTTPS connections?

There is no upstream DoH support in Unbound yet, only downstream. If upstream DoH will be added at a later stage it will benefit from the changes in the PR.

**Review by ralphdolmans — APPROVED** [review]
Looks good to me!

**Comment by iz8mbw** [comment]
Ready to merge? 😁

**Comment by wcawijngaards** [comment]
The branch is ready for use, it is waiting for review(s), before merge.

File: services/outside_network.c
Comment (gthess) [inline services/outside_network.c]: Verbosity number check

File: services/outside_network.c
Comment (gthess) [inline services/outside_network.c]: Verbosity number check

File: services/outside_network.c
Comment (gthess) [inline services/outside_network.c]: The two `if(pend->c->tcp_write_and_read)` can be combined

File: services/outside_network.c
Comment (gthess) [inline services/outside_network.c:2037]: Just a note: this last resort seems less random to me but since we only need the ID for differentiating the queries in the same TCP connection I don't see a problem.

File: services/outside_network.h
Comment (gthess) [inline services/outside_network.h]: The comment about the key is not very clear.

File: testdata/ssl_reuse.tdir/ssl_reuse.conf
Comment (gthess) [inline testdata/ssl_reuse.tdir/ssl_reuse.conf]: Could use the `tls-upstream` option instead.

File: testdata/ssl_reuse.tdir/ssl_reuse.test
Comment (gthess) [inline testdata/ssl_reuse.tdir/ssl_reuse.test]: Mismatch between print value and command.

File: testdata/tcp_reuse.tdir/tcp_reuse.test
Comment (gthess) [inline testdata/tcp_reuse.tdir/tcp_reuse.test]: Mismatch between print value and command.

**Review by gthess — CHANGES_REQUESTED** [review]
Looks good to me also!

I have a question and some nits.
What is the benefit of `MAX_REUSE_TCP_QUERIES`?

The nits:
1. Shouldn't `log_reuse_tcp(<number>, ...)` turn to `log_reuse_tcp(<level>, ...)` and
  `verbose(<number>, ...)` turn to `verbose(<level>, ...)`? Unless they are for dev debugging and will be removed?
2. The `ssl_reuse.tdir` and the `ssl_reuse.*` files therein could be renamed to tls_reuse.

**Comment by wcawijngaards** [comment]
The `MAX_REUSE_TCP_QUERIES` is so that it does not make too many queries at the same time to a destination.  This is because it starts to queue up, but then in terms of TCP buffers, and also, if the connection fails for some reason that is a lot of queries.  So I thought it best not to have too many at the same time on one stream, at the level of thousands it should remove overhead costs just fine.

**Comment by iz8mbw** [comment]
Hi. A my personal note, if I have correctly understood, the connection timeout of reuse is hard-coded to 30 seconds.
This means that the ESTHABILISHED TCP connection to forwarder remains ESTHABILISHED up to 30 seconds (if no new DNS query are made within these 30 seconds).
I think 30 seconds is a low value because if no DNS queries are made for 30 seconds the reuse is useless because unbound will go to create new connections to forwarder.
Can be this value increased? I think 60 seconds make more sense.
Thank you.

File: services/outside_network.c
Comment (wcawijngaards) [reply services/outside_network.c]: fixed.

File: services/outside_network.c
Comment (wcawijngaards) [reply services/outside_network.c]: fixed.

File: services/outside_network.c
Comment (wcawijngaards) [reply services/outside_network.c]: okay.

File: services/outside_network.c
Comment (wcawijngaards) [reply services/outside_network.c:2037]: It is the same randomness? I thought?  Because it picks a random item from the list of available items?

File: services/outside_network.h
Comment (wcawijngaards) [reply services/outside_network.h]: Updated it.

File: testdata/ssl_reuse.tdir/ssl_reuse.conf
Comment (wcawijngaards) [reply testdata/ssl_reuse.tdir/ssl_reuse.conf]: fixed.

File: testdata/ssl_reuse.tdir/ssl_reuse.test
Comment (wcawijngaards) [reply testdata/ssl_reuse.tdir/ssl_reuse.test]: fixed.

File: testdata/tcp_reuse.tdir/tcp_reuse.test
Comment (wcawijngaards) [reply testdata/tcp_reuse.tdir/tcp_reuse.test]: fixed.

**Comment by wcawijngaards** [comment]
Sure, increased the reuse time to 60 seconds in https://github.com/NLnetLabs/unbound/pull/283/commits/6b97cb1dd127125f99ac44df62bfcfc07b5fe1c3 .

File: services/outside_network.c
Comment (gthess) [reply services/outside_network.c:2037]: Upon closer reading this looks good!

**Review by gthess — APPROVED** [review]
Thanks for incorporating the feedback, looks good to me!

**Comment by pengelana** [comment]
Unbound 1.12.0 with patch from #283 and `--enable-tfo-client` & `--enable-tfo-server` will produce error:

`error: read (in tcp s): Transport endpoint is not connected for xxx.xxx.xxx.xxx port 53`

~~~
dig @127.0.0.1 fedoraproject.org

;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 40656
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;fedoraproject.org.             IN      A

;; Query time: 213 msec
~~~

**Comment by wcawijngaards** [comment]
The fix is in commit https://github.com/NLnetLabs/unbound/commit/0502ab3026af59673b21133eb7c5a72126b37d66 . Thanks for the report!

**Pull request merged**
